Assign function keys using descending Z-order.

rogerl50 opened the pull request:
Sort widgets under mouse so that widget on top gets F1 and function keys are assigned in descending order. Fixes #401

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1228/pr-test (or any other room on that server)

ArnoldSmith86 reviewed (approved):
I like it. Maybe we should show cards at the end? And maybe also invisible widgets (so at least a foreign `owner`)?

It's still an improvement either way, so we can also merge as-is.

Just FYI as it's pretty much a question of taste: it would be a bit shorter like this:

`hoveredWidgets.sort((w1,w2)=>w2.calculateZ() - w1.calculateZ());`

96LawDawg commented:
Is it possible to have a trigger to switch and sort in reverse z order.  I like what is being done here as it will provide some consistency.  But sometimes, I want to work on the things at the bottom.  It would be nice if there were a way to trigger a reverse z sort of widgets under the cursor.  Just a suggestion/request.

rogerl50 commented:
> Is it possible to have a trigger to switch and sort in reverse z order. I like what is being done here as it will provide some consistency. But sometimes, I want to work on the things at the bottom. It would be nice if there were a way to trigger a reverse z sort of widgets under the cursor. Just a suggestion/request.

Do you have ideas for what the UI for that might look like?

96LawDawg commented:
> Do you have ideas for what the UI for that might look like?

I was thinking one of two things.  Put a button at the top of the JSON Editor that acts as a toggle (right up there with the other permanent buttons).  Or make a hot key for it (like shift), so when you hold down shift and mouse over, the z order - function key link is reversed.

rogerl50 commented:
> I like it. Maybe we should show cards at the end? And maybe also invisible widgets (so at least a foreign `owner`)?
> 
These both seem reasonable. What test would you propose for "invisible widgets"?

rogerl50 commented:
@ArnoldSmith86 @robartsd Comments on this revision?

RaphaelAlvez commented:
Wait what is the current situation?

rogerl50 commented:
> Wait what is the current situation?

I believe they are assigned in the order in which the widgets appear in the widget array.

ArnoldSmith86 commented:
> @ArnoldSmith86 @robartsd Comments on this revision?

Very nice. Technically you should also check if any parent has the class `foreign` but that's not the most important thing in the world.

rogerl50 commented:
> > @ArnoldSmith86 @robartsd Comments on this revision?
> 
> Very nice. Technically you should also check if any parent has the class `foreign` but that's not the most important thing in the world.

Just to make sure I understand: even if a widget does not have that class, it will be hidden (and therefore should come at the end) if one of its ancestors does? I assume that holds whether or not it is physically enclosed by its ancestor?

ArnoldSmith86 commented:
Yes. Simply because HTML works that way. If an element is not displayed, all its children aren't displayed either.

rogerl50 commented:
> Yes. Simply because HTML works that way. If an element is not displayed, all its children aren't displayed either.

So is there a way to test for this other than iterating up the `parent` chain?

ArnoldSmith86 commented:
I don't know any other way.

rogerl50 commented:
Now widgets whose parents are hidden are assigned F-keys as if they themselves were hidden.

ArnoldSmith86 commented:
:+1:

rogerl50 commented:
> Is it possible to have a trigger to switch and sort in reverse z order. I like what is being done here as it will provide some consistency. But sometimes, I want to work on the things at the bottom. It would be nice if there were a way to trigger a reverse z sort of widgets under the cursor. Just a suggestion/request.

I think I will merge this as is, without addressing your comment, mostly because it's not clear what the right interface would be. Perhaps you should open an issue for this enhancement.
